File: src/main/java/com/kasisoft/libs/common/constants/FileSize.java
package com.kasisoft.libs.common.constants;

/**
 * A collection of file sizes. Calculation won't work on {@link #TerraByte} due to value limits.
 *
 * @author daniel.kasmeroglu@kasisoft.com
 */
public enum FileSize {

    Byte(1L, 1, "B", "B"),
    KiloByte(1000L, 1024, "KB", "KiB"),
    MegaByte(1000L * 1000, 1024L * 1024, "MB", "MiB"),
    GigaByte(1000L * 1000 * 1000, 1024L * 1024 * 1024, "GB", "GiB"),
    TerraByte(0L, 0L, "TB", "TiB");

    private long   humanSize;
    private long   computerSize;
    private String humanUnit;
    private String computerUnit;

    FileSize(long human, long computer, String humanU, String computerU) {
        humanSize    = human;
        computerSize = computer;
        humanUnit    = humanU;
        computerUnit = computerU;
    }

    public String getHumanUnit() {
        return humanUnit;
    }

    public String getComputerUnit() {
        return computerUnit;
    }

    public long getHumanSize() {
        return humanSize;
    }

    public long getComputerSize() {
        return computerSize;
    }

    public long humanSize(int count) {
        return count * humanSize;
    }

    public long computerSize(int count) {
        return count * computerSize;
    }

    public String humanFormat(int count) {
        return "%d %s".formatted(humanSize(count), humanUnit);
    }

    public String computerFormat(int count) {
        return "%d %s".formatted(computerSize(count), computerUnit);
    }

    public FileSize next() {
        return switch (this) {
            case Byte -> KiloByte;
            case KiloByte -> MegaByte;
            case MegaByte -> GigaByte;
            case GigaByte -> TerraByte;
            default -> null;
        };
    }

    public FileSize previous() {
        return switch (this) {
            case KiloByte -> Byte;
            case MegaByte -> KiloByte;
            case GigaByte -> MegaByte;
            case TerraByte -> GigaByte;
            default -> null;
        };
    }

} /* ENDENUM */

File: src/main/kotlin/com/kasisoft/libs/common/constants/FileSize.kt
package com.kasisoft.libs.common.constants

/**
 * A collection of file sizes. Calculation won't work on {@link #TerraByte} due to value limits.
 *
 * @author daniel.kasmeroglu@kasisoft.net
 */
enum class FileSize (val humanSize: Long, val computerSize: Long, val humanUnit: String, val computerUnit: String) {
    
  Byte(1, 1, "B", "B"),
  KiloByte(1000, 1024, "KB", "KiB"),
  MegaByte(1000 * 1000, 1024 * 1024, "MB", "MiB"),
  GigaByte(1000 * 1000 * 1000, 1024 * 1024 * 1024, "GB", "GiB"),
  TerraByte(0L, 0L, "TB", "TiB");

  fun humanSize(count: Int) = count * humanSize
 
  fun computerSize(count: Int) = count * computerSize
  
  fun humanFormat(count: Int) = String.format("%d %s", humanSize(count), humanUnit)

  fun computerFormat(count: Int) = String.format("%d %s", computerSize(count), computerUnit)

  fun next(): FileSize? {
    when (this) {
      Byte -> return KiloByte
      KiloByte -> return MegaByte
      MegaByte -> return GigaByte
      GigaByte -> return TerraByte
      else -> return null
    }
  }

  fun previous(): FileSize? {
    when (this) {
      KiloByte -> return Byte
      MegaByte -> return KiloByte
      GigaByte -> return MegaByte
      TerraByte -> return GigaByte
      else -> return null
    }
  }
    
} /* ENDENUM */